### PR TITLE
[flutter_tools] initialize frontend server with build

### LINKED ISSUE #40388: InitializeFromDill does not work for Android/macOS builds
This slightly slows down the initialization of hot reload on these devices. Ultimately this is due to the app.dill not being in the expected location due to the other build systems (Gradle, assemble) placing it in a different directory.

Longer-term, a better solution would be to re-arrange the resident runner to only run in prebuilt mode, and have the path to the app.dill provided as part of the prebuilt arguments.

### LINKED ISSUE #43464: test/integration.shard/hot_reload_test.dart flakiness
https://cirrus-ci.com/task/4742482515984384?command=main#L159:
```
01:51 +3 -1: test/integration.shard/hot_reload_test.dart: newly added code executes during hot reload [E]                                                                                              
  Expected: contains '(((((RELOAD WORKED)))))'
    Actual: '[{"event":"app.progress","params":{"appId":"506a800b-0dc4-41b0-b02e-ca4fe19c0ad0","id":"1","progressId":"hot.reload","message":"Initializing hot reload..."}}]\n'
              '[{"event":"app.progress","params":{"appId":"506a800b-0dc4-41b0-b02e-ca4fe19c0ad0","id":"2","progressId":null,"message":"Syncing files to device Flutter test device..."}}]\n'
              '[{"event":"app.progress","params":{"appId":"506a800b-0dc4-41b0-b02e-ca4fe19c0ad0","id":"2","progressId":null,"finished":true}}]\n'
              '((((TICK 2))))\n'
              '[{"event":"app.progress","params":{"appId":"506a800b-0dc4-41b0-b02e-ca4fe19c0ad0","id":"1","progressId":"hot.reload","finished":true}}]\n'
              'Reloaded 0 of 474 libraries in 215ms.\n'
              '[{"id":1,"result":{"code":0,"message":"Reloaded 0 of 474 libraries"}}]\n'
              ''
  package:test_api                                                      expect
  org-dartlang-app:///test/integration.shard/hot_reload_test.dart 45:7  main.<fn>
```

cc @DanTup

### LINKED ISSUE #49865: Syncing code takes a long time
## Steps to Reproduce

1. `flutter create repro`
1.  `cd repro`
1. Start an iOS Simulator
1. `flutter run`

=>

```console
Launching lib/main.dart on iPhone 11 Pro Max in debug mode...
Running Xcode build...

...
Syncing files to device iPhone 11 Pro Max...
 7,356ms (!)
```

Using a MacBookPro with Flutter 1.12.13+hotfix.7.

## Logs

```
[  +90 ms] Syncing files to device iPhone 11 Pro Max...
[   +1 ms] Scanning asset files
[   +2 ms] <- reset
[        ] Compiling dart to kernel with 0 updated files
[  +12 ms] /Users/mit/dev/flutter/bin/cache/dart-sdk/bin/dart
/Users/mit/dev/flutter/bin/cache/artifacts/engine/darwin-x64/frontend_server.dart.snapshot --sdk-root
/Users/mit/dev/flutter/bin/cache/artifacts/engine/common/flutter_patched_sdk/ --incremental --target=flutter
-Ddart.developer.causal_async_stacks=true --output-dill /var/folders/yn/6j8s1t215278rzzj2f6kn79c00768h/T/flutter_tool.VOIhPI/app.dill --packages
/Users/mit/tmp/repro1/.packages -Ddart.vm.profile=false -Ddart.vm.product=false
--bytecode-options=source-positions,local-var-info,debugger-stops,instance-field-initializers,keep-unreachable-code,avoid-closure-call-instructions
--enable-asserts --track-widget-creation --filesystem-scheme org-dartlang-root
[  +13 ms] <- compile package:repro1/main.dart
[ +297 ms] [DEVICE LOG] 2020-01-31 09:42:59.944394+0100  localhost Runner[44375]: (RunningBoardServices) Created Activity ID: 0x9e7335,
Description: didChangeInheritances
[+6272 ms] Updating files
[  +90 ms] DevFS: Sync finished
[        ] Syncing files to device iPhone 11 Pro Max... (completed in 6,691ms, longer than expected)
[   +1 ms] Synced 0.9MB.
[   +1 ms] Sending to VM service: _flutter.listViews({})
[   +1 ms] Result: {type: FlutterViewList, views: [{type: FlutterView, id: _flutterView/0x7ffd3f830020, isolate: {type: @Isolate, fixedId: true,
id: isolates/1770924315133767, name: main.dart$main-1770924315133767, number: 1770924315133767}}]}
[        ] <- accept
[        ] Connected to _flutterView/0x7ffd3f830020.
[   +1 ms] 🔥  To hot reload changes while running, press "r". To hot restart (and rebuild state), press "R".
[        ] An Observatory debugger and profiler on iPhone 11 Pro Max is available at: http://127.0.0.1:62107/6c91Ns_CMLM=/
[        ] For a more detailed help message, press "h". To detach, press "d"; to quit, press "q".
```

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

Initialize the frontend_server while performing the platform initialization step (xcode/gradle). Massive performance improvement on startup. Also fixes flakiness in integration tests triggered by changing elapsed time.

Fixes https://github.com/flutter/flutter/issues/40388
Fixes https://github.com/flutter/flutter/issues/49865
Fixes https://github.com/flutter/flutter/issues/43464 (at least I think...)

### CODE REVIEW

**Comment by jonahwilliams** [comment]
This also makes the fast-start workflow less bad, yay!

File: packages/flutter_tools/lib/src/run_hot.dart [hunk 1]
```diff
@@ -329,14 +330,36 @@ class HotRunner extends ResidentRunner {
 
     firstBuildTime = DateTime.now();
 
+    final List<Future<bool>> startupTasks = <Future<bool>>[];
     for (final FlutterDevice device in flutterDevices) {
-      final int result = await device.runHot(
+      // Here we initialize the frontend_server conccurently with the gradle
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/run_hot.dart]: nit. `conccurently` -> `concurrently`

**Review by christopherfujino — APPROVED** [review]
lgtm, with one spelling nit. cool!

File: packages/flutter_tools/lib/src/run_hot.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/run_hot.dart]: Done

**Comment by jonahwilliams** [comment]
Ahh, it started timing out again. Looks like I didn't hit all of the race conditions in the test

**Comment by jonahwilliams** [comment]
Removing unnecessary synchronization for the second tick seems to have removed the flakiness

**Comment by christopherfujino** [comment]
looks gooder

**Pull request merged**